### PR TITLE
Friendly errors

### PR DESCRIPTION
**Pull request by utku-caglayan**

### CODE REVIEW

File: internal/connect.go [hunk 1]
```diff
@@ -33,16 +33,22 @@ func CallClusterOperation(config *hazelcast.Config, operation string, state *str
 	var err error
 	obj := NewRESTCall(config, operation, state)
 	params := obj.params
-	url := obj.url
+	urlStr := obj.url
```
Comment (yuce) [inline internal/connect.go:36]: Let's try to use short variable names if possible

File: internal/links.go [hunk 2]
```diff
@@ -0,0 +1,7 @@
+package internal
+
+const (
+	RESTApiDocs = "https://docs.hazelcast.com/imdg/latest/clients/rest.html"
+	//todo figure out a solution for version in the url
```
Comment (yuce) [inline internal/links.go:5]: `latest` instead of `4.2` works. But we don't need these constants.

File: internal/errors.go [hunk 3]
```diff
@@ -18,15 +18,74 @@ package internal
...
+)
+
+// Cluster errors
+var (
+	ErrRestAPIDisabled = errors.New("Cannot access Hazelcast REST API. Is it enabled? Check this link to find out more: " + RESTApiDocs)
+)
+
+// Network errors
+var (
+	ErrConnectionTimeout = errors.New("Can not connect to Hazelcast Cluster. Please make sure Hazelcast cluster is reachable, up and running. Check this link to create a IMDG cluster: " + QuickStartDocs)
```
Comment (yuce) [inline internal/errors.go:41]: Go error messages should start with lower cased letters (that doesn't apply to application error messages)! But I don't think we need to define these errors.

File: internal/errors.go [hunk 4]
```diff
@@ -18,15 +18,74 @@ package internal
...
+		if netErr.Timeout() {
+			return ErrConnectionTimeout, true
+		}
+
+		var addrErr *net.AddrError
+		if errors.As(err, &addrErr) {
+			return fmt.Errorf("%v:%s", ErrInvalidAddress, addrErr.Error()), true
+		}
+	}
+
```
Comment (yuce) [inline internal/errors.go:73]: Unnecessary blank line

File: internal/errors.go [hunk 5]
```diff
@@ -18,15 +18,74 @@ package internal
...
+	var netOpErr *net.OpError
+	if errors.As(err, &netOpErr) {
+		if netOpErr.Op == "dial" {
+			return ErrConnectionRefused, true
+		} else if netOpErr.Op == "read" {
+			return ErrConnectionRefused, true
+		}
+	}
+
```
Comment (yuce) [inline internal/errors.go:82]: Unnecessary blank line

File: internal/errors.go [hunk 6]
```diff
@@ -18,15 +18,74 @@ package internal
...
+	var syscallErr syscall.Errno
+	//TODO these syscall errors seem platform specific, decide on what to do
+	if errors.As(err, &syscallErr) && syscallErr == syscall.ECONNREFUSED {
+		return ErrConnectionRefused, true
+	}
+
```
Comment (yuce) [inline internal/errors.go:88]: Unnecessary blank line

File: internal/links.go [hunk 7]
```diff
@@ -0,0 +1,7 @@
+package internal
+
+const (
+	RESTApiDocs = "https://docs.hazelcast.com/imdg/latest/clients/rest.html"
```
Comment (yuce) [inline internal/links.go:4]: Initialisms should be all uppercase: Api -> API

File: internal/errors.go [hunk 8]
```diff
@@ -18,15 +18,74 @@ package internal
...
+			return fmt.Errorf("%v:%s", ErrInvalidAddress, addrErr.Error()), true
+		}
+	}
+
+	var netOpErr *net.OpError
+	if errors.As(err, &netOpErr) {
+		if netOpErr.Op == "dial" {
+			return ErrConnectionRefused, true
+		} else if netOpErr.Op == "read" {
+			return ErrConnectionRefused, true
```
Comment (yuce) [inline internal/errors.go:79]: This one doesn't seem like a connection error.

**Review by yuce — CHANGES_REQUESTED** [review]

File: internal/connect.go (hunk 1 above)
Comment (utku-caglayan) [reply internal/connect.go:36]: it conflicts with url package, instead I thought changing the variable name would be better

Comment (yuce) [reply internal/connect.go:36]: OK, that makes sense.

**Review by yuce — APPROVED** [review]
Thanks, looks good!

**Pull request merged**